main.go

```diff
@@ -1,39 +1,64 @@
 package gorillaz
 
 import (
+	"errors"
 	"fmt"
 	"github.com/gorilla/mux"
 	"github.com/skysoft-atm/gorillaz/stream"
 	"github.com/spf13/viper"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/balancer/roundrobin"
 	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/resolver"
 	"net"
 	"net/http"
+	"strconv"
+	"strings"
 	"time"
 )
 
 var initialized = false
 
 type Gaz struct {
-	Router *mux.Router
+	Router             *mux.Router
+	ServiceDiscovery   ServiceDiscovery
+	StreamProviderPort int
 	// use int32 because sync.atomic package doesn't support boolean out of the box
 	isReady        *int32
 	isLive         *int32
 	grpcServer     *grpc.Server
 	streamRegistry *streamRegistry
+	httpListener   *net.Listener
```

Comment:
Do we need both the port and the listener?

Reply:
Listener is an interface, so you don't need a pointer to it

Reply:
the gRPC health check is working, so I will remove the listener and port from the Gaz object